### PR TITLE
e-sparqlEndpoint supports paged queries

### LINKED ISSUE #241: e-sparqlEndpoint should support paged queries
e-sparqlEndpoint should support splitting queries via paging (LIMIT/OFFSET) to enable extraction of more RDF triples than the maximum number that is allowed per single request (e.g., via ResultSetMaxRows in virtuoso.ini). For example, the maximum number or triples per single request is set to 10000 by default in Virtuoso.

When you want to extract more triples, you to do one of two things. If you control the SPARQL endpoint you extract data from, then you can increase the setting for the maximum number of triples for a single request. If you don't control the SPARQL endpoint you extract data from, you need to split your query into multiple queries that retrieve parts of the complete results. Splitting can be done via paging by LIMIT and OFFSET. Since paging requires stable sort order, queries should be wrapped in subqueries to enable to use "scrollable cursors" (see mff-uk/DPUs#78).

A rough Clojure implementation of a SPARQL extractor supporting paged queries based on Apache Jena can be found at <https://github.com/opendatacz/sparql-unlimited>.

An alternative to the LIMIT/OFFSET approach can use the last seen ID (URI), as explained [here](http://use-the-index-luke.com/no-offset) (mentioned by @hanecak [here](https://team.eea.sk/wiki/pages/viewpage.action?pageId=110463376&focusedCommentId=118391471#comment-118391471)).

### PR DESCRIPTION
**Pull request by mkl-eea-sk**
Resolves #241

### CODE REVIEW

**Comment by skrchnavy** [comment]
Dependency on PR UnifiedViews/Core#489

**Comment by tomas-knap** [comment]
1. The combo box with size of the pages can be modified in the DPU configuration. However, if I insert there e.g. 2000 and save and re-open the dpu instance configuration dialog, my value is lost. So it should either support custom values or not. If the value may be customized, user also should not be able to insert there non-integer, e.g., 1000abc - please check the validation. 
2. SparqlEndpoint.dialog.chunksize.description is missing in SK resources (already added to EN resources). Also please consider talking rather about paging, not chunks with respect to the new combo box (see as I adjusted descriptions in EN resources)
3 I added user doc (EN) and changelog. Please add SK user doc (based on EN user doc)

**Comment by tomas-knap** [comment]
When I run the following query: 

```
PREFIX dbo:  <http://dbpedia.org/ontology/>

CONSTRUCT {
  ?village a dbo:Village .
}
WHERE {
  ?village a dbo:Village .
}
```

using the e-sparqlEndpoint before the change, I got: 

screen shot 2015-08-19 at 15 08 04

So it extracts 10 000 data due to limit of the target sparql endpoint, which is correct. 

If I repeat the same query using the update in this pull request and specifying "No chunks", the execution ends with error: 

screen shot 2015-08-19 at 15 13 02

File: e-sparqlEndpoint/src/main/java/eu/unifiedviews/plugins/extractor/sparqlendpoint/SparqlEndpoint.java [hunk 1]
```diff
@@ -82,19 +82,49 @@ protected void innerExecute() throws DPUException {
...
-                long counter = 0;
-                while (result.hasNext()) {
-                    final Statement st = result.next();
-                    // Add to out output.
-                    output.add(st.getSubject(), st.getPredicate(), st.getObject());
-                    // Print info.
-                    ++counter;
-                    if (counter % 100000 == 0) {
-                        LOG.info("{} triples extracted", counter);
+                if (config.getChunkSize() == null) {
```
Comment (tomas-knap) [inline e-sparqlEndpoint/src/main/java/eu/unifiedviews/plugins/extractor/sparqlendpoint/SparqlEndpoint.java]: This is probably not correct, if I save the "No Chunk" option from the configdialog, chunkSize is not null but -1, but it should also follow the if branch

Comment (mkl-eea-sk) [inline e-sparqlEndpoint/src/main/java/eu/unifiedviews/plugins/extractor/sparqlendpoint/SparqlEndpoint.java]: Precisely. But the error is also in Dialog , it should had set chunkSize to null.

**Comment by mkl-eea-sk** [comment]
BTW, the DPU has wrong artifactId :(

**Comment by mkl-eea-sk** [comment]
I have resolved @tomas-knap  comments. Please re-test/

**Comment by tomas-knap** [comment]
Thanks! The problems above are solved. 

Further issues: 
1) If I specify in the original query LIMIT, it is thrown away silently during rewriting. We should at least mark this to the doc or log warning when LIMIT is in the initial query. 
2) If I specify in the original query ORDER BY DESC(?village), it is then automatically rewritten with ORDER BY ASC(?village) - again we should warn user or at least document in the documentation such behavior.
3) If I specify GROUP BY in the query, then the rewriting of the query failed: 

screen shot 2015-08-21 at 11 59 47

Again it may be solved by just checking that GROUP By is not there or just writing to the doc that rewriting this kind of queries is not supported in current version.

So please solve the issues above by warning the user or at least documenting in the DPU doc.

**Comment by tomas-knap** [comment]
Then you can merge without further approval from me.

**Comment by mkl-eea-sk** [comment]
I would be pleased if @jindrichmynarz will read your last comment and comment on this. Since 1. 2. are requirements and to 3. you did not provide an example which I can test (and add to junit tests).

**Comment by jindrichmynarz** [comment]
Ad 1: When there is a `LIMIT` in the outermost query, then it is fine that it is replaced, but it would be preferable if the user is warned about this as @tomas-knap suggests. However, when `LIMIT` is present in an inner query, then it must be preserved. For example:

```
PREFIX dbo: <http://dbpedia.org/ontology/>

CONSTRUCT {
  ?s ?p ?o .
}
WHERE {
  {
    SELECT DISTINCT ?s
    WHERE {
       ?s a dbo:Village .
    }
    LIMIT 500
  }
  ?s ?p ?o .
}
```

Needs to be rewritten to:

```
PREFIX dbo:  <http://dbpedia.org/ontology/>

CONSTRUCT {
  ?s ?p ?o .
}
WHERE {
  {
    SELECT ?s ?p ?o
    WHERE {
      {
         SELECT ?s
         WHERE {
            ?s a dbo:Village .
         }
         LIMIT 500
      }
      ?s ?p ?o .
    }
    ORDER BY ?s ?p ?o
  }
}
LIMIT 10000
OFFSET 0
```

That is, the inner `LIMIT 500` needs to be preserved. The same thing must hold for `ORDER BY` as well. I.e. outermost `ORDER BY` will be discarded (user will be warned), while inner `ORDER BY` must be preserved.

Ad 3) You need to explicitly name all projected variables instead of using `SELECT *`, because `*` is not supported for queries with `GROUP BY`. To test this, we can modify the query 4 to retrieve everything about villages located in more than 1 country. This query:

```
PREFIX dbo: <http://dbpedia.org/ontology/>

CONSTRUCT {
  ?village ?p ?o . 
}
WHERE {
  {
    SELECT ?village
    WHERE {
      GRAPH <http://dbpedia.org> {
        ?village a dbo:Village ;
          dbo:country ?country .
      }
    }
    GROUP BY ?village
    HAVING (COUNT(DISTINCT ?country) > 1)
  }
  ?village ?p ?o .
}
```

Needs to be rewritten as:

```
PREFIX dbo: <http://dbpedia.org/ontology/>

CONSTRUCT {
  ?village ?p ?o . 
}
WHERE {
  {
    SELECT ?village ?p ?o
    WHERE {
      {
        SELECT ?village
        WHERE {
          GRAPH <http://dbpedia.org> {
            ?village a dbo:Village ;
              dbo:country ?country .
          }
        }
        GROUP BY ?village
        HAVING (COUNT(DISTINCT ?country) > 1)
      }
      ?village ?p ?o .
    }
    ORDER BY ?village ?p ?o
  }
}
LIMIT 10000
OFFSET 0
```

**Comment by tomas-knap** [comment]
@jindrichmynarz Thanks for adding particular examples!

**Comment by mkl-eea-sk** [comment]
This pull request has its codebase moved to: https://github.com/UnifiedViews/Plugins/pull/315
since this pull request has been reverted in develop by https://github.com/UnifiedViews/Plugins/commit/a7dc440d40eb397792314b36a1f064f1434ca522 and https://github.com/UnifiedViews/Plugins/commit/b9d95d3d9eeb9a35ff12b6c8e0c26bff39c24768

**Pull request merged**